infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/UniqueWorkspacePVCStrategy.java

```diff
@@ -106,42 +125,47 @@ public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
       ephemeralWorkspaceAdapter.provision(k8sEnv, identity);
       return;
     }
+
     LOG.debug("Provisioning PVC strategy for workspace '{}'", workspaceId);
+
+    fillInExistingPVCs(k8sEnv, workspaceId);
```

Comment:
General (optional) suggestion: `provision` would be more clear if it didn't mix cluster PVCs and env PVCs. The current process is
1. Add existing PVCs to env
2. Figure out which existing PVC corresponds to the user-defined one, if it exists
3. Replace user defined PVC with existing PVC, if it exists.

Since all of these PVCs are stored in the same map, it is hard to reason about, especially across many method calls (i.e. what is the current state of `k8sEnv.getPersistentVolumeClaims()`?). If instead clusterPVCs and envPVCs were kept separate, with the last step being clearing envPVCs and replacing them with clusterPVCs, it would be more clear IMO.

Would it be possible to instead e.g.
1. Get cluster PVCs
2. For each user-defined PVC, get matching cluster PVC (or create one if does not exist) -- this is what's done in groupByVolumeName, but it operates on the mixed map instead of separate ones).
3. Replace user-defined PVCs with cluster PVCs, add projects/logs/etc PVCs to env.
4. Do necessary rewriting in pods, etc.

I know this probably isn't as clear as it could be, but my goal would be to make explicit the process of matching user-defined PVCs to provisioned PVCs.

Reply:
Thanks for the deep analysis of PVCs strategies and well-described proposal. Will try.

Reply:
Generally, I like that idea but I doubt it will not be simple. Like

> 4. Do necessary rewriting in pods, etc.

Needs information from the second step. 

@amisevsk Are you OK with doing this refactoring in another PR to finally get this reviewed changes merged and make further review easier?

Reply:
Proposed refactoring is done in the changes that are going to be in my next PR https://github.com/sleshchenko/che/commit/23fa7d5017af9ef690704cd5123e508e99892b92

I've shared it just for preview, some test adaptation, cleaning up the code, java docs adding may still be needed there.

Reply:
:+1: Agree that it's not necessary for this PR.